src/python/grpcio/grpc/_channel.py

```diff
@@ -314,70 +363,111 @@ def add_callback(self, callback):
+    existence of any other thread, including the "channel spin thread".
+    However, this means that its interface is entirely synchronous. So this
+    class cannot fulfill the grpc.Future interface.
+    """
+
+    def initial_metadata(self):
+        """See grpc.Call.initial_metadata"""
+        with self._state.condition:
+            # NOTE(gnossen): Based on our initial call batch, we are guaranteed
+            # to receive initial metadata before any messages.
+            while self._state.initial_metadata is None:
+                self._get_next_event()
             return self._state.initial_metadata
 
     def trailing_metadata(self):
         """See grpc.Call.trailing_metadata"""
         with self._state.condition:
-
-            def _done():
-                return self._state.trailing_metadata is not None
-
-            _common.wait(self._state.condition.wait, _done)
+            if self._state.trailing_metadata is None:
+                raise RuntimeError(
+                    "Cannot get trailing metadata until RPC is completed.")
             return self._state.trailing_metadata
 
-    # TODO(https://github.com/grpc/grpc/issues/20763): Drive RPC progress using
-    # the calling thread.
     def code(self):
         """See grpc.Call.code"""
         with self._state.condition:
-
-            def _done():
-                return self._state.code is not None
-
-            _common.wait(self._state.condition.wait, _done)
+            if self._state.code is None:
+                raise RuntimeError("Cannot get code until RPC is completed.")
             return self._state.code
 
     def details(self):
         """See grpc.Call.details"""
         with self._state.condition:
-
-            def _done():
-                return self._state.details is not None
-
-            _common.wait(self._state.condition.wait, _done)
+            if self._state.details is None:
+                raise RuntimeError("Cannot get details until RPC is completed.")
             return _common.decode(self._state.details)
 
-    def _next(self):
+    def _get_next_event(self):
+        event = self._call.next_event()
         with self._state.condition:
-            if self._state.code is None:
-                operating = self._call.operate(
-                    (cygrpc.ReceiveMessageOperation(_EMPTY_FLAGS),), None)
-                if operating:
-                    self._state.due.add(cygrpc.OperationType.receive_message)
-            elif self._state.code is grpc.StatusCode.OK:
-                raise StopIteration()
-            else:
-                raise self
+            callbacks = _handle_event(event, self._state,
+                                      self._response_deserializer)
+            for callback in callbacks:
+                # NOTE(gnossen): We intentionally allow exceptions to bubble up
+                # to the user when running on a single thread.
+                callback()
+        return event
```

Comment:
That's true at the moment. This is sort of in preparation for the refactor I mentioned this morning. I'd really like there to be a `_get_next_message` method instead of relying on the shared state of `self._state` everywhere in the code, with zero abstraction. `_get_next_event()` is an ingredient in that. At the moment, `_consume_next_event()` is probably a better representation of what it's doing (unfortunately). If you feel strongly about it, I'll change it. Otherwise, I'd like to leave it as is for now.

Reply:
Discussed offline. Switching to `_consume_next_event()` until such time as the aforementioned refactor actually materializes.